WalletWasabi/EventSourcing/InMemoryEventRepository.cs

```diff
@@ -0,0 +1,243 @@
+			= ImmutableList<string>.Empty;
+
+		private static readonly IComparer<WrappedEvent> WrappedEventSequenceIdComparer
+			= Comparer<WrappedEvent>.Create((a, b) => a.SequenceId.CompareTo(b.SequenceId));
+
+		private ConcurrentDictionary
+			// aggregateType
+			<string,
+			ConcurrentDictionary
+				// aggregateId
+				<string,
+				(
+					// SequenceId of the last event of this aggregate
+					long TailSequenceId,
+
+					// Ordered list of events
+					ImmutableList<WrappedEvent> Events
+				)>> AggregatesEventsBatches
+		{ get; } = new();
+
+		private ConcurrentDictionary
+			// aggregateType
+			<string,
+			(
+				// Index of the last aggregateId in this aggregateType
+				long TailIndex,
+
+				// List of aggregate Ids in this aggregateType
+				ImmutableSortedSet<string> Ids
+			)> AggregatesIds
+		{ get; } = new();
+
+		/// <inheritdoc/>
+		public Task AppendEventsAsync(
+			string aggregateType,
+			string aggregateId,
+			IEnumerable<WrappedEvent> wrappedEvents)
+		{
+			Guard.NotNullOrEmpty(nameof(aggregateType), aggregateType);
+			Guard.NotNullOrEmpty(nameof(aggregateId), aggregateId);
+			Guard.NotNull(nameof(wrappedEvents), wrappedEvents);
+
+			var wrappedEventsList = wrappedEvents.ToList().AsReadOnly();
+
+			if (wrappedEventsList.Count == 0)
+			{
+				return Task.CompletedTask;
+			}
+			var firstSequenceId = wrappedEventsList[0].SequenceId;
+			var lastSequenceId = wrappedEventsList[^1].SequenceId;
+
+			if (firstSequenceId <= 0)
+			{
+				throw new ArgumentException("First event sequenceId is not natural number.", nameof(wrappedEvents));
+			}
+
+			if (lastSequenceId <= 0)
+			{
+				throw new ArgumentException("Last event sequenceId is not a positive integer.", nameof(wrappedEvents));
+			}
+
+			if (lastSequenceId - firstSequenceId + 1 != wrappedEventsList.Count)
+			{
+				throw new ArgumentException("Event sequence ids are inconsistent.", nameof(wrappedEvents));
+			}
+
+			var aggregateEventsBatches = AggregatesEventsBatches.GetOrAdd(aggregateType, _ => new());
+			var (tailSequenceId, events) = aggregateEventsBatches.GetOrAdd(
+				aggregateId,
+				_ => (0, ImmutableList<WrappedEvent>.Empty));
+
+			if (tailSequenceId + 1 < firstSequenceId)
+			{
+				throw new ArgumentException($"Invalid firstSequenceId (gap in sequence ids) expected: '{tailSequenceId + 1}' given: '{firstSequenceId}'.", nameof(wrappedEvents));
+			}
+
+			// no action
+			Validated();
+
+			var newEvents = events.AddRange(wrappedEventsList);
+
+			// Atomically detect conflict and replace lastSequenceId and lock to ensure strong order in eventsBatches.
+			if (!aggregateEventsBatches.TryUpdate(
+				key: aggregateId,
+				newValue: (lastSequenceId, newEvents),
+				comparisonValue: (firstSequenceId - 1, events)))
+			{
+				Conflicted(); // no action
+				throw new OptimisticConcurrencyException(
+					$"Conflict while committing events. Retry command. aggregate: '{aggregateType}' id: '{aggregateId}'");
```

Comment:
@molnard Isn't it better to have a slightly longer line than format it like this? 

If anything, it makes regex fulltext searches harder.